### DISCUSSION TITLE
Hot spares and how it related to the "autoreplace" zpool property is confusing. Do you need that set to "on" in order for hot spares to work as expected?

### DISCUSSION THREAD

**Discussion by recklessnl** [openzfs/zfs#12865] [Q&A]
From the docs:

> autoreplace=on|off
    Controls automatic device replacement. If set to off, device replacement must be initiated by the administrator by using the zpool replace command. If set to on, any new device, found in the same physical location as a device that previously belonged to the pool, is automatically formatted and replaced. The default behavior is off. This property can also be referred to by its shortened column name, replace. Autoreplace can also be used with virtual disks (like device mapper) provided that you use the /dev/disk/by-vdev paths setup by vdev_id.conf. See the vdev_id(8) manual page for more details. Autoreplace and autoonline require the ZFS Event Daemon be configured and running. See the zed(8) manual page for more details.

But this doesn't list hot spares anyway. I think the docs should be clarified here. This property defaults to off - will hot spares automatically replace a failed drive with this set to off? I would assume they would (otherwise, what's the point of a hot spare?) but I'd rather make sure.

**Comment by rincebrain** [comment]
(Disclaimer: I do not and have not, to my recollection, ever used either feature, so if the documentation and reality disagree, I will be wrong in the same way.)

Autoreplace is a property for, AIUI, "if you stick a disk in what ZFS can tell is the same physical location as a failed disk, ZFS will act like you ran `zpool replace pool olddisk newdisk`" - that is, imitating what some large proprietary disk solutions do.

Hot spares will automatically be `zpool attach`ed (more or less) if a disk faults and there are any available in the pool, and have nothing to do with that property; they just are treated as being temporary replacements unless you tell it to make it permanent by `zpool detach`ing the faulted disk.

If you can think of a way you'd like to see the docs clarified, I'd encourage you to open a PR - since you're one of the people who finds it insufficient,  I think you're better positioned to contribute a more useful explanation for people who likewise find it unclear than people who have been using ZFS for a long time and take all of its longstanding behaviors for granted. (For example, I'd probably just add a single line in each of `zpoolconcepts.7` and `zpoolprops.7` about how they're not at all related, but I have no good sense of whether that'd be sufficient...)
